File: vscode-extension/src/lsp/ast.ts
export interface Span {
    start: number;
    end: number;
}

export interface StructField {
    name: string;
    fieldType: string | null;
    typeAnnotation: string | null;
}

export interface CompGenerator {
    var: string;
    iter: Expr;
    condition: Expr | null;
}

export interface TableFieldPair {
    key: Expr;
    value: Expr;
}

export interface TableFieldValue {
    value: Expr;
}

export type TableField = TableFieldPair | TableFieldValue;

export type Expr =
    | { kind: "Number"; value: number }
    | { kind: "Str"; value: string }
    | { kind: "FString"; value: string }
    | { kind: "Bool"; value: boolean }
    | { kind: "Nil" }
    | { kind: "Ident"; name: string }
    | { kind: "SelfExpr" }
    | { kind: "UnaryMinus"; inner: Expr }
    | { kind: "Grouping"; inner: Expr }
    | { kind: "Array"; elements: Expr[] }
    | { kind: "Table"; fields: TableField[] }
    | { kind: "Index"; obj: Expr; index: Expr }
    | { kind: "Call"; func: string; args: Expr[] }
    | { kind: "MethodCall"; obj: Expr; field: string; isColon: boolean; args: Expr[] }
    | { kind: "Member"; obj: Expr; field: string; isColon: boolean }
    | { kind: "Binary"; left: Expr; op: string; right: Expr }
    | { kind: "Ternary"; cond: Expr; thenExpr: Expr; elseExpr: Expr }
    | { kind: "Logical"; left: Expr; op: string; right: Expr }
    | { kind: "Not"; inner: Expr }
    | { kind: "Function"; params: string[]; block: Stmt[] }
    | { kind: "AwaitExpr"; inner: Expr }
    | { kind: "ListComp"; elt: Expr; generators: CompGenerator[] };

export type Stmt =
    | { kind: "Local"; names: string[]; value: Expr | null; access: string; span: Span; typeAnnotations: (string | null)[] }
    | { kind: "Assign"; target: Expr; value: Expr; op: string | null; span: Span }
    | { kind: "Return"; value: Expr | null; span: Span }
    | { kind: "If"; cond: Expr; thenBlock: Stmt[]; elseIfBlocks: [Expr, Stmt[]][]; elseBlock: Stmt[] | null; span: Span }
    | { kind: "While"; cond: Expr; block: Stmt[]; span: Span }
    | { kind: "For"; vars: string[]; iter: Expr; block: Stmt[]; span: Span; typeAnnotations: (string | null)[] }
    | { kind: "FuncDef"; name: string; params: string[]; paramTypes: (string | null)[]; paramDefaults: (Expr | null)[]; block: Stmt[]; access: string; isAsync: boolean; span: Span; returnType: string | null }
    | { kind: "ClassDef"; name: string; body: Stmt[]; access: string; span: Span; extends: string | null }
    | { kind: "ExprStmt"; expr: Expr; span: Span }
    | { kind: "EnumDef"; name: string; variants: string[]; access: string; span: Span }
    | { kind: "StructDef"; name: string; fields: StructField[]; access: string; span: Span }
    | { kind: "Import"; path: string; alias: string; span: Span }
    | { kind: "Break"; span: Span }
    | { kind: "Continue"; span: Span }
    | { kind: "TryCatch"; tryBlock: Stmt[]; catchClauses: [string | null, string | null, Stmt[]][]; finallyBlock: Stmt[] | null; span: Span }
    | { kind: "DecoratedStmt"; decorators: string[]; stmt: Stmt; span: Span };

File: src/ast.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Default)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

impl Span {
    pub fn new(start: usize, end: usize) -> Self {
        Self { start, end }
    }
}

// ==========================================
// 2. THE AST
// ==========================================
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct StructField {
    pub name: String,
    pub field_type: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct CompGenerator {
    pub var: String,
    pub iter: Expr,
    pub condition: Option<Expr>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum TableField {
    Pair { key: Expr, value: Expr },
    Value(Expr),
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum Expr {
    Number(f64),
    Str(String),
    FString(String),
    Bool(bool),
    Nil,
    Ident(String),
    SelfExpr,
    UnaryMinus(Box<Expr>),
    Grouping(Box<Expr>),
    Array(Vec<Expr>),
    Table(Vec<TableField>),
    Index {
        obj: Box<Expr>,
        index: Box<Expr>,
    },
    Call {
        func: String,
        args: Vec<Expr>,
    },
    MethodCall {
        obj: Box<Expr>,
        field: String,
        is_colon: bool,
        args: Vec<Expr>,
    },
    Member {
        obj: Box<Expr>,
        field: String,
        is_colon: bool,
    },
    Binary {
        left: Box<Expr>,
        op: String,
        right: Box<Expr>,
    },
    Ternary {
        cond: Box<Expr>,
        then_expr: Box<Expr>,
        else_expr: Box<Expr>,
    },
    Logical {
        left: Box<Expr>,
        op: String,
        right: Box<Expr>,
    },
    Not(Box<Expr>),
    Function {
        params: Vec<String>,
        block: Vec<Stmt>,
    },
    AwaitExpr(Box<Expr>),
    ListComp {
        elt: Box<Expr>,
        generators: Vec<CompGenerator>,
    },
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum Stmt {
    Local {
        name: String,
        value: Option<Expr>,
        access: String,
        type_annotation: Option<String>,
        #[serde(skip)]
        span: Span,
    },
    Assign {
        target: Expr,
        value: Expr,
        op: Option<String>,
        #[serde(skip)]
        span: Span,
    },
    Return {
        value: Option<Expr>,
        #[serde(skip)]
        span: Span,
    },
    If {
        cond: Expr,
        then_block: Vec<Stmt>,
        else_if_blocks: Vec<(Expr, Vec<Stmt>)>,
        else_block: Option<Vec<Stmt>>,
        #[serde(skip)]
        span: Span,
    },
    While {
        cond: Expr,
        block: Vec<Stmt>,
        #[serde(skip)]
        span: Span,
    },
    For {
        var: String,
        iter: Expr,
        block: Vec<Stmt>,
        type_annotation: Option<String>,
        #[serde(skip)]
        span: Span,
    },
    FuncDef {
        name: String,
        params: Vec<String>,
        param_types: Vec<Option<String>>,
        param_defaults: Vec<Option<Expr>>,
        block: Vec<Stmt>,
        access: String,
        is_async: bool,
        return_type: Option<String>,
        #[serde(skip)]
        span: Span,
    },
    ClassDef {
        name: String,
        body: Vec<Stmt>,
        access: String,
        #[serde(skip)]
        span: Span,
    },
    ExprStmt {
        expr: Expr,
        #[serde(skip)]
        span: Span,
    },
    EnumDef {
        name: String,
        variants: Vec<String>,
        access: String,
        #[serde(skip)]
        span: Span,
    },
    StructDef {
        name: String,
        fields: Vec<StructField>,
        access: String,
        #[serde(skip)]
        span: Span,
    },
    Import {
        path: String,
        alias: String,
        #[serde(skip)]
        span: Span,
    },
    Break {
        #[serde(skip)]
        span: Span,
    },
    Continue {
        #[serde(skip)]
        span: Span,
    },
    TryCatch {
        try_block: Vec<Stmt>,
        catch_clauses: Vec<(Option<String>, Option<String>, Vec<Stmt>)>,
        finally_block: Option<Vec<Stmt>>,
        #[serde(skip)]
        span: Span,
    },
    DecoratedStmt {
        decorators: Vec<String>,
        stmt: Box<Stmt>,
        #[serde(skip)]
        span: Span,
    },
}
